### DISCUSSION TITLE
zig怎么样读取注册表的？

### DISCUSSION THREAD

**Discussion by qcjxberin** [zigcc/forum#99] [新手提问]
```c
#include <windows.h>
#include <stdio.h>

int main() {
    SetConsoleOutputCP(CP_UTF8);  // 中文乱码

    HKEY hKey;
    LONG lResult;
    DWORD dwType = REG_SZ;  // 设置为字符串类型
    char szData[1024];
    DWORD dwSize = sizeof(szData);

    // 创建或打开键
    //lResult = RegCreateKeyEx(HKEY_CURRENT_USER,  // 预定义的根键
    //                         "Software\\TestApp",  // 子键路径
    //                        0, // 保留
    //                         NULL, // 类字符串
    //                         REG_OPTION_NON_VOLATILE,  // 保存信息
    //                         KEY_WRITE, // 安全访问掩码
    //                         NULL, // 安全属性
    //                         &hKey, // 输出的键句柄
    //                         NULL); // 处理新旧值

    // 打开键
    lResult = RegOpenKeyEx(HKEY_CURRENT_USER,
                           "Software\\Tencent\\WeChat",
                           0,
                           KEY_READ,
                           &hKey);

    if (lResult != ERROR_SUCCESS) {
        printf("打开键失败\n");
        return 1;
    }

    // 读取键
    lResult = RegQueryValueEx(hKey,
                              "InstallPath",
                              NULL,
                              &dwType,
                              (LPBYTE)szData,
                              &dwSize);

    if (lResult != ERROR_SUCCESS) {
        printf("读取键值失败\n");
        RegCloseKey(hKey);
        return 1;
    }

    printf("成功读取注册表：%s\n", szData);

    // 关闭键
    RegCloseKey(hKey);

    return 0;
}
```
这是我用c实现的代码，用zig实现在转换时总有报错，麻烦看看怎么样实现zig读取注册表的代码。

**Comment by jiacai2050** [comment]
最好把『操作的步骤』和『报错信息』一起发上来，还有操作系统版本。

**Comment by jinzhongjia** [comment]
查看 [`std.os.windows.advapi32`](https://ziglang.org/documentation/master/std/#A;std:os.windows.advapi32), 注册表管理，服务管理，和用户账号管理的api都在里面，但zig对windows api支持不全（不打算支持全部api，太多了）

**Reply by jinzhongjia** [reply]
`RegOpenKeyExW`
`RegQueryValueExW`
`RegCloseKey`
 这三个 api zig 都有提供

**Reply by qcjxberin** [reply]
是否有具体的Demo代码，目前试过各种方式也问过AI，无解

**Comment by qcjxberin** [comment]
@jiacai2050 

const std = @import("std");
const windows = std.os.windows;

pub fn main() !void {
    const allocator = std.heap.page_allocator;
    const key_path = "Software\\Tencent\\WeChat";
    //const value_name = "InstallPath"; // 你需要读取的具体值名称

    // 将 UTF-8 字符串转换为 UTF-16 LE
    var key_path_buffer: [256]u16 = undefined;
    const key_path_utf16 = try std.unicode.utf8ToUtf16Le(&key_path_buffer, key_path);
    defer allocator.free(key_path_utf16);

    // var value_name_buffer: [256]u16 = undefined;
    // const value_name_utf16 = try std.unicode.utf8ToUtf16Le(&value_name_buffer, value_name);
    // defer allocator.free(value_name_utf16);

    // 打开注册表键
    var hkey: windows.HKEY = undefined;
    const result = windows.advapi32.RegOpenKeyExW(windows.HKEY_CURRENT_USER, key_path_buffer[0..key_path_utf16 :0], 0, windows.KEY_READ, &hkey);
    if (result != 0) {
        return error.OperatingSystemError;
    }
    defer {
        const close_result = windows.advapi32.RegCloseKey(hkey);
        if (close_result != 0) {
            std.log.warn("Failed to close registry key: {}", .{close_result});
        }
    }
}
![image](https://github.com/zigcc/forum/assets/5792218/35d83b44-bd4d-4100-9b28-c48bf603440b)
版本0.12  windows11系统
